src/uproot/_dask.py

```diff
@@ -971,11 +971,111 @@ def load_buffers(
 
         return container
 
+    def load_virtual_buffers(
+        self,
+        tree: HasBranches,
+        keys: frozenset[str],
+        start: int,
+        stop: int,
+        decompression_executor,
+        interpretation_executor,
+        options: Any,
+    ) -> Mapping[str, AwkArray]:
+        awkward = uproot.extras.awkward()
+
+        def generator(tree, buffer_key):
+            form_key, attribute = self.parse_buffer_key(buffer_key)
+            key = self._form_key_to_key[form_key]
+            branch = tree[key]
+
+            def _generator():
+                array = branch.array(
+                    entry_start=start,
+                    entry_stop=stop,
+                    interpretation_executor=interpretation_executor,
+                    decompression_executor=decompression_executor,
+                    library="ak",
+                    ak_add_doc=options.get("ak_add_doc"),
+                )
+
+                # add to access_log
+                access_log = options.get("access_log")
+                if access_log is not None:
+                    if not hasattr(access_log, "__iadd__"):
+                        raise ValueError(
+                            f"{access_log=} needs to implement '__iadd__'."
+                        )
+                    else:
+                        access_log += [Accessed(branch=key, buffer_key=buffer_key)]
+
+                # Convert the sub-array into buffers
+                ttree_subform, _, ttree_container = awkward.to_buffers(array)
+
+                # Load the associated projection subform
+                projection_subform = self._form.content(key)
+
+                # Correlate each TTree form key with the projection form key
+                for (src, src_dtype), (dst, dst_dtype) in zip(
+                    ttree_subform.expected_from_buffers().items(),
+                    projection_subform.expected_from_buffers(self.buffer_key).items(),
+                ):
+                    # Return the corresponding array from the TTree if buffer key matches
+                    if buffer_key == dst:
+                        if src_dtype != dst_dtype:
+                            raise TypeError(
+                                f"Data type mismatch: {src_dtype} != {dst_dtype}"
+                            )
+                        return ttree_container[src]
+
+                # Raise an error if the buffer key is not found
+                raise ValueError(
+                    f"Buffer key {buffer_key} not found in form {self._form}"
+                )
+
+            return _generator
+
+        container = {}
+        for buffer_key, _ in self._form.expected_from_buffers().items():
+            container[buffer_key] = generator(tree, buffer_key)
+
+        return container
+
+
+def form_with_unique_keys(form: Form, key: str) -> Form:
+    awkward = uproot.extras.awkward()
```

Comment:
If this function is performance-sensitive or called multiple times, moving this import/lookup to the module scope would be better.
Or at least as a default parameter to avoid repeated calls.

Reply:
oh, wait - apart from that line this function is a copy of `form_with_unique_keys`: https://github.com/dask-contrib/dask-awkward/blob/3440ce07c2c2e14657b1e48b42f4437b6d7e7849/src/dask_awkward/lib/utils.py#L140

Why?

Reply:
It's not performance sensitive, it's just called once per `.virtual_arrays()` call.

> oh, wait - apart from that line this function is a copy of form_with_unique_keys

Because there should not be a dependency on dask-awkward here. 
It's an unfortunate duplicate right now, but it would need a whole more rethinking and restructuring the whole form-manipulation logic in dask-awkward and uproot.dask. I'm happy to look into this (I planned to anyway), but I did not want to block this feature from being added for now.

Reply:
I would assume the name `_dask` implies the dependency on `dask_awkward`. Besides, there is still other dependency in this module: https://github.com/scikit-hep/uproot5/blob/56eebf6fdd44f01b66d91b3c6acf346dcecf1f6b/src/uproot/_dask.py#L1228 

Would using the `utils` from `dask_awkward` block the new feature? Thanks.

Reply:
There's no dependency on dask-awkward: 
It's only imported when `uproot.extras.dask_awkward()` is called, and those are always inside a function and not at the top module level. The line you've linked is never called in `.virtual_arrays()`.